File: example/src/androidTest/java/eu/darken/mvpbakery/example/screens/MainActivityTest.java
package eu.darken.mvpbakery.example.screens;


import android.app.Activity;
import android.support.test.InstrumentationRegistry;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;
import android.support.v4.app.Fragment;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import dagger.android.AndroidInjector;
import eu.darken.mvpbakery.example.ExampleApplicationMock;
import eu.darken.mvpbakery.example.R;
import eu.darken.mvpbakery.example.screens.counting.CountingComponent;
import eu.darken.mvpbakery.example.screens.counting.CountingFragment;
import eu.darken.mvpbakery.example.screens.counting.CountingPresenter;
import eu.darken.mvpbakery.injection.ComponentSource;
import eu.darken.mvpbakery.injection.ManualInjector;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.when;

@RunWith(AndroidJUnit4.class)
public class MainActivityTest {

    @Rule public MockitoRule mockitoRule = MockitoJUnit.rule();

    @Rule public ActivityTestRule<MainActivity> activityRule = new ActivityTestRule<>(MainActivity.class, true, false);

    ExampleApplicationMock app;


    @Mock MainPresenter mainPresenter;
    @Mock MainComponent mainComponent;

    @Mock ComponentSource<Fragment> fragmentInjector;
    @Mock CountingPresenter countingPresenter;
    @Mock CountingComponent countingComponent;

    @Before
    public void setUp() {
        app = (ExampleApplicationMock) InstrumentationRegistry.getTargetContext().getApplicationContext();
        app.setActivityComponentSource(new ActivityInjector());

        doAnswer(invocation -> {
            MainActivity mainActivity = invocation.getArgument(0);
            mainActivity.componentSource = fragmentInjector;
            return null;
        }).when(mainComponent).inject(any());
        when(mainComponent.getPresenter()).thenReturn(mainPresenter);
        when(mainPresenter.getComponent()).thenReturn(mainComponent);


        doAnswer(invocation -> {
            // Nothing to inject atm
            return null;
        }).when(countingComponent).inject(any());
        when(fragmentInjector.get(any())).then(invocation -> countingComponent);
        when(countingComponent.getPresenter()).thenReturn(countingPresenter);
        when(countingPresenter.getComponent()).thenReturn(countingComponent);
    }

    public class ActivityInjector implements ManualInjector<Activity> {

        @Override
        public AndroidInjector get(Activity instance) {
            return mainComponent;
        }

        @Override
        public void inject(Activity instance) {
            mainComponent.inject((MainActivity) instance);
        }

    }

    @Test
    public void checkFragmentShowing() throws Throwable {
        activityRule.launchActivity(null);

        activityRule.runOnUiThread(() -> activityRule.getActivity().showFragment(CountingFragment.class));
        onView(withId(R.id.fragment_button)).check(matches(withText("+1")));
    }

    @Test
    public void checkCounter() throws Throwable {
        activityRule.launchActivity(null);

        onView(withId(R.id.bindcounter)).check(matches(withText("Activity Rotation Count: 0")));

        activityRule.runOnUiThread(() -> activityRule.getActivity().showBinderCounter(1));
        onView(withId(R.id.bindcounter)).check(matches(withText("Activity Rotation Count: 1")));

        activityRule.runOnUiThread(() -> activityRule.getActivity().showBinderCounter(2));
        onView(withId(R.id.bindcounter)).check(matches(withText("Activity Rotation Count: 2")));
    }
}

File: example/src/androidTest/java/eu/darken/mvpbakery/example/screens/MainActivityTest.kt
package eu.darken.mvpbakery.example.screens


import android.app.Activity
import androidx.fragment.app.Fragment
import androidx.test.core.app.ApplicationProvider
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import androidx.test.rule.ActivityTestRule
import dagger.android.AndroidInjector
import eu.darken.mvpbakery.example.ExampleApplicationMock
import eu.darken.mvpbakery.example.R
import eu.darken.mvpbakery.example.screens.counting.CountingComponent
import eu.darken.mvpbakery.example.screens.counting.CountingFragment
import eu.darken.mvpbakery.example.screens.counting.CountingPresenter
import eu.darken.mvpbakery.injection.ComponentSource
import eu.darken.mvpbakery.injection.ManualInjector
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import org.junit.Before
import org.junit.Rule
import org.junit.Test

class MainActivityTest {
    @get:Rule
    var activityRule = ActivityTestRule(MainActivity::class.java, true, false)

    private lateinit var app: ExampleApplicationMock

    @MockK
    lateinit var mainPresenter: MainPresenter
    @MockK
    lateinit var mainComponent: MainComponent

    @MockK
    lateinit var fragmentInjector: ComponentSource<Fragment>
    @MockK
    lateinit var countingPresenter: CountingPresenter
    @MockK
    lateinit var countingComponent: CountingComponent

    @Before
    fun setUp() {
        MockKAnnotations.init(this, relaxUnitFun = true)
        app = ApplicationProvider.getApplicationContext()
        app.setActivityComponentSource(ActivityInjector())

        every { mainComponent.inject(any()) } answers {
            val mainActivity = firstArg<MainActivity>()
            mainActivity.componentSource = fragmentInjector
            mainActivity.presenter = mainPresenter
        }

        every { mainComponent.presenter } returns (mainPresenter)
        every { mainPresenter.component } returns (mainComponent)

        every { countingComponent.inject(any()) } answers { }
        @Suppress("UNCHECKED_CAST")
        every { fragmentInjector[any()] } returns (countingComponent as AndroidInjector<Fragment>)
        every { countingComponent.presenter } returns (countingPresenter)
        every { countingPresenter.component } returns (countingComponent)
    }

    inner class ActivityInjector : ManualInjector<Activity> {
        @Suppress("UNCHECKED_CAST")
        override fun get(instance: Activity): AndroidInjector<Activity> = mainComponent as AndroidInjector<Activity>

        override fun inject(instance: Activity) = mainComponent.inject(instance as MainActivity)
    }

    @Test
    fun checkFragmentShowing() {
        activityRule.launchActivity(null)

        activityRule.runOnUiThread { activityRule.activity.addFragment(CountingFragment::class.java) }

        onView(withId(R.id.fragment_button)).check(matches(withText("+1")))
    }

    @Test
    fun checkCounter() {
        activityRule.launchActivity(null)

        onView(withId(R.id.bindcounter)).check(matches(withText("Activity Rotation Count: 0")))

        activityRule.runOnUiThread { activityRule.activity.showBinderCounter(1) }
        onView(withId(R.id.bindcounter)).check(matches(withText("Activity Rotation Count: 1")))

        activityRule.runOnUiThread { activityRule.activity.showBinderCounter(2) }
        onView(withId(R.id.bindcounter)).check(matches(withText("Activity Rotation Count: 2")))
    }
}
